File: ios/Methods/DeviceInformationChangeMethodHandler.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

struct DeviceInformationChangeMethodHandler: MethodHandler {
	func execute(using client: MobileAuthenticationClient, with message: ChannelInMessage) throws {
		let message: DeviceInformationChangeMessage = try validate(message: message)
		let operation = DeviceInformationChangeOperation(operationId: message.operationId)
		OperationCache.shared.put(operation, using: message.operationId)

		let deviceInformationChange = client.operations.deviceInformationChange
		if let name = message.name {
			deviceInformationChange.name(name)
		}

		if let fcmRegistrationToken = message.fcmRegistrationToken, !fcmRegistrationToken.isEmpty {
			deviceInformationChange.fcmRegistrationToken(fcmRegistrationToken)
		}

		if let disablePushNotifications = message.disablePushNotifications, disablePushNotifications {
			deviceInformationChange.disablePushNotifications()
		}
		
		if let retryPolicy = message.retryPolicy {
			deviceInformationChange.retryPolicy(retryPolicy.wrapped)
		}

		if let requestHeaders = message.requestHeaders {
			deviceInformationChange.requestHeaders(requestHeaders.namesAndValues)
		}

		if message.onSuccessProvided {
			deviceInformationChange.onSuccess(operation.onSuccess)
		}

		if message.onErrorProvided {
			deviceInformationChange.onError(operation.onError)
		}

		deviceInformationChange.execute()
	}
}

File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/methods/DeviceInformationChangeMethodHandler.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.methods

import ch.nevis.mobile.authentication.sdk.react.cache.OperationCache
import ch.nevis.mobile.authentication.sdk.react.cache.operation.DeviceInformationChangeOperation
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.ChannelInMessage
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.DeviceInformationChangeMessage
import ch.nevis.mobile.sdk.api.MobileAuthenticationClient

class DeviceInformationChangeMethodHandler : MethodHandler() {
    override fun execute(
        client: MobileAuthenticationClient,
        channelInMessage: ChannelInMessage,
    ) {
        val message = validate<DeviceInformationChangeMessage>(channelInMessage)
        val operation = DeviceInformationChangeOperation(message.operationId)
        OperationCache.createOperation(message.operationId, operation)

        val deviceInformationChange = client.operations().deviceInformationChange()
        message.name?.let {
            deviceInformationChange.name(it)
        }

        message.fcmRegistrationToken?.let {
            deviceInformationChange.name(it)
        }

        message.disablePushNotifications?.let {
            if (it) {
                deviceInformationChange.disablePushNotifications()
            }
        }

        message.retryPolicy?.let {
            deviceInformationChange.retryPolicy(it)
        }

        message.requestHeaders?.let {
            deviceInformationChange.requestHeaders(it)
        }

        if (message.onSuccessProvided) {
            deviceInformationChange.onSuccess {
                operation.onSuccess(null)
            }
        }

        if (message.onErrorProvided) {
            deviceInformationChange.onError(operation.onError)
        }

        deviceInformationChange.execute()
    }
}
